### PR TITLE
Add PGM support to load_image/save_image

### PR DESCRIPTION
**Pull request by dpalermo**
- Added PGM support for reading/writing graymaps
- Improve performance of reading/writing PPM pixmaps
- Allow for building in environments that don't have libpng (define -DNOPNG)
- Allow for building in environments without full C++11 STL (define -DNOSTL11)

### CODE REVIEW

**Comment by zvookin** [comment]
Can you please do this without introducing any C preprocessor macro definitions? If you absolutely must, they should be prefixed with something like "HALIDE_" to avoid name collisions. Glancing over this, I see no reason to introduce any however.

**Comment by dpalermo** [comment]
The NOPNG macro controls whether the header includes png.h as well as the function uses of the same.  The additional command line flag -DNOPNG only needs to be passed when building.If a build environment didn't have libpng (e.g. Hexagon standalone, Android NDK).  I don't know how to detect this automatically.

The NOSTL11 macro could be removed if you are OK with always replacing x.data() with &x[0] in the STL vectors.

Let me know if the following are OK with you and I'll make the changes:
  - renaming NOPNG to HALIDE_NOPNG
  - dropping NOSTL11 and always using &x[0]

**Comment by steven-johnson** [comment]
What environments don't support proper C++11 STL? Halide has required C++11
for a good while now.

**Comment by dpalermo** [comment]
The Hexagon LLVM C++ compiler supports C++11, but the STL (from a third party) in the Hexagon SDK currently does not provide the C++11 introduced data() accessor method.  We have already reported this missing feature and it is being corrected so it should be available in a release later this year.

**Comment by dpalermo** [comment]
Changed NOPNG to HALIDE_NOPNG
Remove NOSTL11, always use &data[0] instead of data.data()

**Comment by dpalermo** [comment]
Simplified patch by removing changes for ASCII PGM/PPM support

File: tools/halide_image_io.h [hunk 1]
```diff
@@ -93,6 +95,8 @@ inline void swap_endian_16(bool little_endian, uint16_t &value) {
     }
 }
 
+#define FILEOPENER_MAX_LINE_LEN 1024
```
Comment (steven-johnson) [inline tools/halide_image_io.h]: Please use constexpr for constants; preprocessor macros are best avoided unless there is no alternative

File: tools/halide_image_io.h [hunk 2]
```diff
@@ -310,6 +332,111 @@ bool save_png(ImageType &im, const std::string &filename) {
...
     return true;
+#endif // HALIDE_NOPNG
+}
+
+template<typename ImageType, Internal::CheckFunc check = Internal::CheckReturn>
+bool load_pgm(const std::string &filename, ImageType *im) {
+
+    /* open file and test for it being a pgm */
+    Internal::FileOpener f(filename.c_str(), "rb");
+    if (!check(f.f != nullptr, "File %s could not be opened for reading\n", filename.c_str())) return false;
```
Comment (steven-johnson) [inline tools/halide_image_io.h]: Please write all if's with enclosing braces on multiplelines:

    if (!check(...)) {
        return false;
    }

File: tools/halide_image_io.h [hunk 3]
```diff
@@ -310,6 +332,111 @@ bool save_png(ImageType &im, const std::string &filename) {
...
+        fmt_binary = true;
+    if (!check(fmt_binary, "Input is not binary PGM\n")) return false;
+
+    f.readLine();
+    if (!check(sscanf(f.buf, "%d %d\n", &width, &height) == 2, "Could not read PGM width and height\n")) return false;
+    f.readLine();
+    if (!check(sscanf(f.buf, "%d", &maxval) == 1, "Could not read PGM max value\n")) return false;
+
+    int bit_depth = 0;
+    if (maxval == 255) { bit_depth = 8; }
```
Comment (steven-johnson) [inline tools/halide_image_io.h]: Fix brace formatting
   
     if () {
     } else if {
     } else {
     }

File: tools/halide_image_io.h [hunk 4]
```diff
@@ -310,6 +332,111 @@ bool save_png(ImageType &im, const std::string &filename) {
...
+    else if (maxval == 65535) { bit_depth = 16; }
+    else { if (!check(false, "Invalid bit depth in PGM\n")) return false; }
+
+    // Graymap
+    *im = ImageType(width, height);
+
+    // convert the data to ImageType::ElemType
+    if (bit_depth == 8) {
+        std::vector<uint8_t> data(width*height);
+        if (!check(fread((void *) &data[0], sizeof(uint8_t), width*height, f.f) == (size_t) (width*height), "Could not read PGM 8-bit data\n")) return false;
```
Comment (steven-johnson) [inline tools/halide_image_io.h]: please prefer C++ style casts

**Comment by dpalermo** [comment]
I will be happy to make the style changes but I was just matching the format used in the existing code.

File: tools/halide_image_io.h [hunk 5]
```diff
@@ -102,9 +106,19 @@ struct FileOpener {
             fclose(f);
         }
     }
+    // read a line of data skipping lines that begin with '#"
+    char *readLine() {
```
Comment (dsharletg) [inline tools/halide_image_io.h]: This function really should take a pointer to a buffer with a size, rather than return a pointer to a member buffer.

**Comment by dsharletg** [comment]
Regarding style issues: we should just run clang-format on this whole file. As you pointed out, the old code had a lot of style issues as well.

**Comment by dpalermo** [comment]
Pass a buffer and size to FileOpener::readLine
Remove member buffer from FileOpener
Note: readline was added to skip comments that may appear in the PGM/PPM header

File: tools/halide_image_io.h [hunk 6]
```diff
@@ -94,6 +96,7 @@ inline void swap_endian_16(bool little_endian, uint16_t &value) {
 }
 
 struct FileOpener {
+    static constexpr int FileOpener_MaxLineLen = 1024;
```
Comment (dsharletg) [inline tools/halide_image_io.h]: This looks unused

File: tools/halide_image_io.h [hunk 7]
```diff
@@ -310,6 +330,112 @@ bool save_png(ImageType &im, const std::string &filename) {
...
+
+    f.readLine(buf, 1024);
+    if (!check(sscanf(buf, "%d %d\n", &width, &height) == 2, "Could not read PGM width and height\n")) return false;
+    f.readLine(buf, 1024);
+    if (!check(sscanf(buf, "%d", &maxval) == 1, "Could not read PGM max value\n")) return false;
+
+    int bit_depth = 0;
+    if (maxval == 255) { bit_depth = 8; }
+    else if (maxval == 65535) { bit_depth = 16; }
+    else { if (!check(false, "Invalid bit depth in PGM\n")) return false; }
```
Comment (dsharletg) [inline tools/halide_image_io.h]: The braces on this line are in a bizarre place, probably intended:

    else if (!check(false, "Invalid bit depth in PGM\n")) { return false; }

**Comment by dpalermo** [comment]
Removed unused constexpr
Moved PGM/PPM invalid bit depth check brace

**Comment by ronlieb** [comment]
Hi Folks, what else is needed to move this patch forward?
we are very dependent on this functionality to get Android to run our toyCamera and dispatch to the HVX (hexagon).

**Comment by dsharletg** [comment]
I'll merge this, and do some cleanup on the code (both old and new) later, the old PPM code needed a lot of work here.

**Pull request merged**